rclcpp_lifecycle/test/test_lifecycle_node.cpp

```diff
@@ -447,6 +447,72 @@ TEST_F(TestDefaultStateMachine, bad_mood) {
   EXPECT_EQ(1u, test_node->number_of_callbacks);
 }
 
+
+TEST_F(TestDefaultStateMachine, shutdown_from_each_primary_state) {
```

Comment:
this test doesn't seem to exercise the new feature, as it calls `shutdown` before letting the node go out of scope.
should we have an additional test where we verify that the shutdown callback is always invoked when the destructor is invoked if the node wasn't in the correct state?
for example the shutdown callback could set a global flag that is checked by the test

Reply:
i think that is hard, `LifecycleNode` derived class `EmptyLifecycleNode` will be destructed before `LifecycleNode`, that means it will not call the `on_shutdown` callback via `LifecycleNode` destruction.

i tried as following.

```cpp
class EmptyLifecycleNode : public rclcpp_lifecycle::LifecycleNode
{
public:
  explicit EmptyLifecycleNode(const std::string & node_name)
  : rclcpp_lifecycle::LifecycleNode(node_name)
  {
    shutdown_cb_called = false;
  }

  inline static bool shutdown_cb_called;

protected:
  rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
  on_shutdown(const rclcpp_lifecycle::State &)
  {
    EXPECT_EQ(State::TRANSITION_STATE_SHUTTINGDOWN, get_current_state().id());
    shutdown_cb_called = true;
    return rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
  }
};
```

```cpp
TEST_F(TestDefaultStateMachine, test_shutdown_on_dror) {
  auto success = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS;
  auto reset_key = rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::ERROR;

  // PRIMARY_STATE_INACTIVE to shutdown
  {
    auto ret = reset_key;
    auto test_node = std::make_shared<EmptyLifecycleNode>("testnode");
    EXPECT_EQ(State::PRIMARY_STATE_UNCONFIGURED, test_node->get_current_state().id());
    auto configured = test_node->configure(ret);
    EXPECT_EQ(success, ret);
    EXPECT_EQ(configured.id(), State::PRIMARY_STATE_INACTIVE);
    EXPECT_FALSE(EmptyLifecycleNode::shutdown_cb_called);
  }
  EXPECT_TRUE(EmptyLifecycleNode::shutdown_cb_called); // this will be false
}
```

Reply:
@alsora what do you think? i think current test codes can verify the situation that destructor can meet to call `shutdown`.

Reply:
Ah, sorry maybe we can use `register_on_shutdown` and free func with global variable to make sure that shutdown is called via destructor. (i guess that originally @Barry-Xu-2018 said.)

Reply:
addressed in https://github.com/ros2/rclcpp/pull/2450/commits/73bb3e5840501de4f08e7b6635589138b74f78c1